### DISCUSSION TITLE
Unstable/sudden  position changes

### DISCUSSION THREAD

**Discussion by tommhuth** [pmndrs/use-cannon#96] [General]
I've made a simple game where the player moves forward at a constant speed and the path the player navigates is automatically generated in front (and  clear behind) the player. I did this with a custom `Cannonjs` implementation based off of React Three Fiber's own example. Curious of the potential performance benefits of using this library, I  decided to swap out my own physics handling.

However, as the player moves forward and the path is generated, the screen will flicker as the position of the simulated bodies seems to change for a single frame (temporarily), as if the syncing of something is temporarily off. It also looks like it is only the meshes that changes, as so far as I can tell, it does not seem to affect the simulated bodies. Check out the following [CodeSandbox](https://codesandbox.io/s/unstable-boxes-5w9vq). Because is happens so quickly, here's two screenshots right before and as it happens:

Before:
<img width="713" alt="Screen Shot 2020-07-10 at 00 54 44" src="https://user-images.githubusercontent.com/1224941/87098897-acc13b00-c248-11ea-98e3-0d2b8d381e6f.png">

Only a couple of frames later:
<img width="733" alt="Screen Shot 2020-07-10 at 00 54 35" src="https://user-images.githubusercontent.com/1224941/87098902-afbc2b80-c248-11ea-8b6c-2f96d73d23b3.png">

A few frames later again:
<img width="713" alt="Screen Shot 2020-07-10 at 01 22 36" src="https://user-images.githubusercontent.com/1224941/87100138-e778a280-c24b-11ea-962b-a0437cef95f7.png">

I can't see anywhere in my code where body position is changed, bodies are only added and removed. Is it the way state is handled thru `Zustand`? My own physics handling did not have this problem (running on the main thread), so could this be a side effect of having the simulation running in a separate thread? Or is there something else I'm missing?

**Comment by codynova** [comment]
My guess would be the state management, but there are a lot of variables here. I'd try to get a minimum viable reproduction of the problem, without the state management etc. and go from there.

**Comment by tommhuth** [comment]
@codynova  I trimmed down the Codesandbox to a minimal one: https://codesandbox.io/s/unstable-boxes2-gg7mr

No `zustand`, just `useState`, where boxes are added and removed randomly at an interval. The jumpy behaviour  persists as in the first example. 

(Removing the `use-cannon` parts fixes the problem.)

**Reply by codynova** [reply]
@drcmda Did you mention that you had seen an issue with `use-cannon` with Body positions on the first frame after they've been added? That seems to be what's happening here

**Reply by drcmda** [reply]
yes, it seems like for a single frame objects appear at 0 position for some reason.

**Reply by tommhuth** [reply]
@codynova  I've also noticed that bug (which my own Cannon implementation had and which I simply solved with setting the mesh position directly on first render only and make sure `position` is supplied to `new Body()` as well), but is this relevant to this because here _all_ meshes' position seem to get jumbled, not just a single new one, and it seems to only happen when something is _removed_? Almost as if there's an index thats off?
